dask_expr/_merge.py

```diff
@@ -314,6 +321,16 @@ def merge_indexed_right(self):
             self.right_index or _contains_index_name(self.right, self.right_on)
         ) and self.right.known_divisions
 
+    def _on_condition_alread_partitioned(self, expr, on):
+        if not isinstance(on, list):
+            result = (
+                on in expr.unique_partition_mapping_columns
+                or (on,) in expr.unique_partition_mapping_columns
+            )
+        else:
+            result = tuple(on) in expr.unique_partition_mapping_columns
+        return result and expr.npartitions == self.npartitions
```

Comment:
IIUC, whether this works depends on the order of `on` and tuples in `unique_partition_mapping_columns`. We should make this order-independent.

Reply:
See https://github.com/coiled/benchmarks/pull/1471#discussion_r1528086910

Reply:
It is already order dependent, but I'll add a test to test for this

Reply:
That would need a bit of care and goes a little into join reordering, so want to keep this out for now